models/product.py

```diff
@@ -375,41 +450,36 @@ def get_by_name(cls, name):
             return None
         return ProductView.query.filter_by(name=name).one_or_none()
 
-    @classmethod
-    def get_by_token(cls, token):
-        if token is None:
-            return None
-        return ProductView.query.filter_by(token=token).one_or_none()
-
-    def is_accessible(self, user, user_token=None):
+    def is_accessible(self, user, voucher=None):
         " Whether this ProductView is accessible to a user."
         if user.is_authenticated and user.has_permission("admin"):
             # Admins always have access
             return True
 
-        if not self.token and datetime.utcnow() < config_date("SALES_START"):
-            # If TICKET_SALES is set, but sales haven't started, restrict access to token views
+        if not self.vouchers and datetime.utcnow() < config_date("SALES_START"):
             return False
 
+        # CfP voucher
         if self.cfp_accepted_only:
-            # Token is an optional override
-            if self.token and user_token == self.token:
+            if user and user.is_authenticated and user.is_cfp_accepted:
                 return True
-
-            if user.is_authenticated and user.is_cfp_accepted:
-                return True
-
             return False
 
-        if self.token:
-            if user_token == self.token:
-                return True
+        if self.vouchers:
+            if not voucher:
+                return False
+
+            for vchr in self.vouchers:
```

Comment:
I think I'd rather this was a query than iterating over all vouchers, as at the moment it's going to load all vouchers in for a productview every time it gets hit.